explorer/interpreter/type_checker.cpp

```diff
@@ -4138,16 +4138,25 @@ auto TypeChecker::TypeCheckPattern(
     }
     *trace_stream_ << "\n";
   }
-  if (require_irrefutable) {
-    switch (p->kind()) {
-      case PatternKind::AutoPattern:
-      case PatternKind::ExpressionPattern:
+  switch (p->kind()) {
+    case PatternKind::AutoPattern:
```

Comment:
To the extent that `auto` is a pattern at all, I think it's irrefutable. This is a pre-existing problem, so I'm fine with leaving this as-is if it's not easy to fix.

Reply:
Can you please clarify, what kind of change are you looking for? We already discussed `auto` as a pattern and I filed #2788 to make it not-a-pattern because it's substantial work. Is this overlapping with #2788?

Reply:
Ah, I see now, you're objecting to the fix from #2782: requiring names where we expect bindings.

Here are a few examples of how that would crop up, and associated bugs:

- `fn F(auto)` -- Expects a type instead of a value to be passed in.
  - `F(i32)` is valid.
  - `F(3)` gives: COMPILATION ERROR: test.carbon:6: mismatch in non-deduced types, `i32` is not implicitly convertible to `type`
- `class C(auto)` -- Same as fn
- `for (auto in x)` -- Doesn't use pattern syntax: "SYNTAX ERROR: test.carbon:9: syntax error, unexpected AUTO, expecting identifier"
- `var auto = 3`, `let` similarly
  - Local (statement): valid
  - Global (declaration): Doesn't use pattern syntax, same error as `for`

I think there needs to be a clear decision on whether `auto` should be a pattern, and whether these syntaxes should be supported. For now, I'm going to leave these as disallowed because I think the introduction of `unused` and advice for `_` from #2022 should be interpreted as saying the above should not be treated as valid syntax.

I'll note this discussion on Discord for others to weigh in, since that's where discussion led to #2788.

Reply:
Repeating a clarification from Discord: I agree that the behavior implemented in #2782 (disallowing `fn F(auto)`, etc) is the behavior that we want. I see now that deleting this line would revert that behavior, so I retract the suggestion.